### ISSUE TITLE
Android ARM64: CPU (NEON/KleidiAI) + QNN NPU + Vulkan (RDNA)

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/any-stt#4]
### NNAPI は非推奨 → QNN / Vulkan / CPU に方針転換

NNAPI は Android 15 (2024) で deprecated。CNN 時代の設計で transformer の ops (LayerNorm, Softmax, MHA) をカバーできず Google が放棄。後継は LiteRT + ベンダー固有アクセラレータだが、GGUF モデルを直接 NPU に流すには自前で ggml QNN backend を書く必要がある。

### ベンダー別 NPU/GPU 状況

| ベンダー | チップ | NPU | GPU | 3rd party NPU アクセス |
|---------|-------|-----|-----|---------------------|
| **Qualcomm** | 8 Gen 3 ~34 TOPS, 8 Elite ~45 TOPS | Hexagon HTP | Adreno | **QNN SDK** (C API, dlopen) |
| **Samsung** | Exynos 2500 ~59 TOPS | 独自 NPU | **Xclipse (AMD RDNA 3.5)** | Neural SDK 非公開。NPU 不可 |
| **Google** | Tensor G5 | Edge TPU | PowerVR (ドライバ壊れ) | experimental SDK のみ。不可 |
| **MediaTek** | Dimensity 9400 APU 890 | APU | Immortalis-G925 | NeuroPilot (将来対応) |

### NPU でのトランスフォーマー実行の制約

**どのモバイル NPU も transformer をエンドツーエンドで実行できない。**
- NPU 得意: INT8/INT4 MatMul (HMX で高速)
- NPU 不可: LayerNorm, Softmax, Attention scaling → CPU フォールバック必須
- FP16 MatMul は NPU 上で CPU INT8 比 **159x 遅い** (Qualcomm 実測)
- GGUF の per-block 量子化 (32要素) が QNN の per-tensor/per-channel にマッピング不可 → カスタム dequant op (LUT) か offline re-quant が必要
- **ヘテロジニアス実行** (MatMul→NPU, LayerNorm/Softmax→CPU) が唯一の実用的アーキテクチャ

### QNN SDK 技術詳細

- Pure C API。`dlopen(libQnnHtp.so)` → `QnnInterface_t` vtable 取得
- Graph ベース: create → addNode → finalize (コンパイル・最適化) → execute
- HTP sub-libs: V73 (8Gen2), V75 (8Gen3), V79 (8Elite), V81 (次世代)
- **1グラフ = 1バックエンド**。ヘテロジニアス実行はグラフ分割で自前実装
- Context caching (`qnn_context.bin`) 必須。finalize が数秒〜十数秒
- メモリ: `rpcmem_alloc` (ION/DMA-buf) + 4K alignment で zero-copy
- **再配布不可**。ユーザーが QNN SDK を別途入手 (CUDA と同じモデル)
- 最新: QAIRT SDK v2.42.0
- **llama.cpp PR #12063** が ggml + QNN 統合の最良リファレンス

### QNN HTP 対応 ops

| Operation | HTP 対応 | 備考 |
|-----------|---------|------|
| MatMul / FullyConnected | ✅ 高速 | HMX で加速。主要ワークロード |
| Conv2d | ✅ 高速 | |
| Softmax | ✅ | LUT ベース (1.26-2.19x) |
| Add / Mul / Scale | ✅ | |
| Transpose / Reshape / Concat | ✅ | |
| Gather / Dequantize | ✅ | LUT ベース |
| GELU | ✅ | activation function として |
| RoPE | ✅ | compute volume 小 |
| **LayerNorm** | ❌ 分解必要 | elementwise ops に分解 |
| **RMSNorm** | ❌ 分解必要 | L2Norm + Mul に分解 |
| 複雑な制御フロー | ❌ | |

### 主要端末 GPU スペック一覧

| SoC | GPU | コア/CU | クロック | FP32 TFLOPS | メモリ帯域 |
|-----|-----|---------|---------|-------------|-----------|
| **SD 8 Elite** | Adreno 830 | 12 CU / 3072 shader | 1100 MHz | ~3.7 | 77 GB/s |
| **SD 8 Gen 3** | Adreno 750 | 1536 shader | 900 MHz | ~2.8 | 51 GB/s |
| **SD 8 Gen 2** | Adreno 740 | 1536 ALU | 680 MHz | ~2.1 | 51 GB/s |
| **Exynos 2600** | Xclipse 960 (RDNA 4) | 16 CU | 985 MHz | ~4-5 | TBD |
| **Exynos 2500** | Xclipse 950 (RDNA 3.5) | **16 CU** | 1300 MHz | **~4.0-4.5** | 77 GB/s |
| **Exynos 2400** | Xclipse 940 (RDNA 3) | 12 CU | 1095 MHz | ~3.4 | 51 GB/s |
| **Dimensity 9400** | Immortalis-G925 | 12 core | 1612 MHz | **~5-8** | 77 GB/s |
| **Dimensity 9300** | Immortalis-G720 | 12 core | 1300 MHz | ~6.0 | 77 GB/s |
| **Tensor G5** | IMG DXT-48-1536 | **2 core** | 1100 MHz | **~1.6** | 51 GB/s |
| **Tensor G4** | Mali-G715 MC6 | 6 core | 940 MHz | ~1.1 | 51 GB/s |
| **Tensor G3** | Mali-G715 MC7 | 7 core | 890 MHz | ~1.2 | 51 GB/s |

### Async Compute と GPU 占有率

**RDNA (Samsung Exynos) は async compute をネイティブサポート。STT compute と UI 描画が並行実行可能。**

| GPU ファミリ | Async Compute | STT との共存 |
|-------------|--------------|-------------|
| **Xclipse / RDNA (Samsung)** | **ネイティブ対応** | compute と graphics が独立パイプライン。低優先度キューに STT を投げて UI は graphics キューで回せる |
| **Immortalis (MediaTek)** | 部分対応 | vertex/compute と fragment が別スケジューリング。fragment と overlap 可能 |
| **Adreno (Qualcomm)** | ほぼ不可 | 統一シェーダ、1 queue family。全 work が同じ実行ユニットを奪い合う |
| **Mali (Pixel 9)** | 部分対応 | compute と fragment の overlap 可。ただし GPU 自体が弱すぎる |
| **PowerVR (Pixel 10)** | 不明 | ドライバが壊れてるので論外 |

#### GPU 占有率を下げる戦略

- **Vulkan queue priority**: compute を低優先度に設定 → UI rendering が常に優先
- **小 dispatch 分割**: 1 巨大 dispatch ではなく多数の小 dispatch → preemption しやすく
- **GPU 利用率 30-50%** が UI jank 回避の実用的な目安
- **Android ADPF** (Dynamic Performance Framework) でサーマル監視 → 熱くなったら compute を抑制

#### Samsung RDNA の async compute が意味すること

README の前提「Android は GPU を 3D レンダリングに使ってるから Vulkan STT は競合する」は **RDNA の async compute では成り立たない**。Samsung Exynos デバイスでは:
- Compute queue に STT MatMul を低優先度で投げる
- Graphics queue は UI 描画に専念
- 両者が HW レベルで並行動作

→ **Samsung 向けは Vulkan (async compute) を auto-select する価値がある**

### デバイス別ベスト戦略 (更新)

| デバイス | 最速パス | 次善 | フォールバック |
|---------|---------|------|-------------|
| **Snapdragon** | QNN NPU (Phase 2) | CPU NEON + KleidiAI | — |
| **Exynos (RDNA GPU)** | **Vulkan GPU (async compute)** | CPU SME2 + KleidiAI | CPU NEON |
| **Dimensity (Immortalis)** | Vulkan GPU (部分 async) | CPU NEON + KleidiAI | — |
| **Pixel 9 (Mali)** | CPU I8MM + KleidiAI | — | — |
| **Pixel 10 (PowerVR)** | CPU I8MM + KleidiAI | — | — |

### kotoba-v2 (large ベース ~1.6GB) 推定性能

| パス | RTF (30s audio) | 実用性 |
|------|----------------|--------|
| QNN NPU (INT8) | ~0.02x | ✅ 最速 |
| Vulkan RDNA (Exynos, ~4 TFLOPS) | ~0.3-0.5x | ✅ 実用的 |
| Vulkan Immortalis (~6 TFLOPS) | ~0.2-0.4x | ✅ 実用的 (要検証) |
| CPU SME2 + Q4_0 | ~0.5-1x | ⚠️ ギリギリ |
| CPU NEON + Q4_0 | ~1-2x | ⚠️ 遅い |
| CPU NEON + FP16 | ~5-10x | ❌ 使い物にならない |

### ggml Vulkan の Android 現状

| GPU | 動作 | 備考 |
|-----|------|------|
| **Adreno (Qualcomm)** | ❌ モデルロード失敗 | |
| **Xclipse / RDNA (Samsung)** | **✅ 好成績** | デスクトップ AMD と同系統。async compute 対応 |
| **Immortalis (MediaTek)** | ⚠️ 未検証 | Mali 系だが Immortalis は別アーキ。要検証 |
| **Mali (Pixel 9)** | ❌ 15x slower than CPU | Mali 対応 "not planned" |
| **PowerVR (Pixel 10)** | ❌ 出力 all-zero | ドライバ壊れ |

### ARM CPU 最適化の状況

| 機能 | Cortex-X4 (Tensor G5) | Cortex-X925 (Exynos 2500) | 次世代 (Lumex) |
|------|----------------------|---------------------------|---------------|
| NEON | ✅ | ✅ | ✅ |
| Dotprod | ✅ | ✅ | ✅ |
| I8MM | ✅ | ✅ | ✅ |
| SVE2 | ✅ | ✅ | ✅ |
| **SME2** | ❌ | **✅** | ✅ |

- **KleidiAI**: ggml に統合済み。DotProd/I8MM/SME2 の優先順で最適カーネル選択
- **SME2**: Exynos 2500 (X925) で利用可能。~55% 高速化 (llama.cpp ベンチ)

### 既存の whisper.cpp Android バックエンド

| パス | 成熟度 | 備考 |
|------|--------|------|
| ggml-cpu (NEON) | ✅ Production | KleidiAI 込みで最も信頼性高い |
| ggml-hexagon | ❌ Abandoned (2025-07) | ~0.5 TOPS (理論値の1%)。dep policy 違反 |
| ggml-vulkan (Android) | ⚠️ RDNA のみ | Mali/Adreno/PowerVR は使い物にならない |

---

### Phase 1: CPU (NEON + KleidiAI) — 全デバイスで動く
- [ ] whisper.cpp を Android NDK (`aarch64-linux-android`) でクロスコンパイル
- [ ] CPU NEON 推論動作確認
- [ ] KleidiAI 有効化 (`GGML_CPU_KLEIDIAI`)
- [ ] Android 上のハードウェア検出 (CPU features, GPU vendor, SoC 判定)
- [ ] `Backend::Nnapi` → `Backend::Qnn` に改名 (NNAPI deprecated のため)

### Phase 2: QNN NPU backend (Snapdragon 向け)
- [ ] ggml QNN backend 自前実装 (llama.cpp PR #12063 参考)
- [ ] QNN SDK の dlopen + QnnInterface_t vtable 取得
- [ ] MatMul / FullyConnected の HTP オフロード
- [ ] GGUF per-block 量子化 → QNN 対応フォーマットへの変換 (LUT dequant or re-quant)
- [ ] ヘテロジニアス実行: MatMul→NPU, LayerNorm/Softmax→CPU
- [ ] Context caching (`qnn_context.bin`) で初回以降の起動高速化
- [ ] Snapdragon 8 Gen 3 / 8 Elite での動作確認

### Phase 3: Vulkan GPU (Samsung RDNA + MediaTek Immortalis)
- [ ] ggml Vulkan backend の Android ビルド有効化
- [ ] Samsung Exynos (Xclipse/RDNA) での動作確認
- [ ] **Async compute 実装**: 低優先度 compute queue に STT dispatch、UI と並行実行
- [ ] MediaTek Immortalis (G720/G925) での Vulkan 動作検証
- [ ] selector.rs: RDNA GPU 検出時に Vulkan を auto-select、Immortalis は検証後に追加
- [ ] Vulkan shader cache の永続化
- [ ] ADPF 統合: サーマル監視で compute 抑制

## Depends on
- #1 Core crate ✅
- #2 Hardware detection ✅
- #3 whisper.cpp FFI ✅

## 参考資料
- [QNN SDK Documentation](https://docs.qualcomm.com/bundle/publicresource/topics/80-63442-2/overview.html)
- [llama.cpp QNN Backend PR #12063](https://github.com/ggml-org/llama.cpp/pull/12063)
- [KleidiAI](https://github.com/ARM-software/kleidiai)
- [ExecuTorch Samsung Exynos Backend](https://github.com/pytorch/executorch)
- [ggml-hexagon (abandoned)](https://github.com/jeffzhou2000/ggml-hexagon) — 参考のみ、依存しない
- [NNAPI Migration Guide](https://developer.android.com/ndk/guides/neuralnetworks/migration-guide)
- [ARM Blog: Async Compute on Mali GPUs](https://developer.arm.com/community/arm-community-blogs/b/mobile-graphics-and-gaming-blog/posts/using-asynchronous-compute-on-arm-mali-gpus)
- [Android ADPF](https://developer.android.com/games/optimize/adpf)
- [Arm SME2 Android](https://www.infoq.com/news/2025/07/arm-sme2-android/)
- [Samsung SME2 Collaboration](https://semiconductor.samsung.com/news-events/tech-blog/expanding-cpu-capabilities-for-on-device-ai-with-arm-sme2/)

**Comment by m96-chan** [comment]
## Phase 3 最終ベンチマーク — CPU+NPU ヘテロジニアス実行

### テスト環境

- **デバイス**: REDMAGIC 9 Pro (NX769J), Snapdragon 8 Gen 3 (SM8650)
- **CPU**: 2× Cortex-X4 + 4× Cortex-A720 + 2× Cortex-A520, 8 cores, OpenMP ON
- **NPU**: Hexagon V75 HTP, QNN SDK v2.26.0
- **Model**: whisper tiny.en (77 MB, F16, 4 encoder layers, n_state=384)
- **Audio**: JFK inaugural (11.0s, 16kHz mono WAV)

### 結果

| Backend | Median | Min | Max | RTF |
|---------|--------|-----|-----|-----|
| **CPU E2E** (whisper.cpp, 8T) | **867ms** | 849ms | 882ms | 0.079 |
| **NPU encoder** (QNN HTP V75) | **103ms** | 57ms | 122ms | 0.009 |
| **Hybrid decode-only** | **148ms** | 113ms | 181ms | 0.013 |
| CPU preprocess | 645ms | 639ms | 647ms | 0.059 |

### ヘテロジニアス合計

**NPU encoder (103ms) + decode-only (148ms) = 251ms**
vs CPU E2E 867ms → **3.5x 高速化**, RTF 0.023

### E2E 出力検証

```
Input:  JFK inaugural address (jfk.wav, 11.0s)
CPU:    "And so my fellow Americans ask not what your country can do for you ask what you can do for your country."
Hybrid: "And so my fellow Americans ask not what your country can do for you ask what you can do for your country."
Result: MATCH ✅
```

### 注意: Vulkan backend 干渉問題

`--features vulkan` でビルドすると `libggml-vulkan.so` が ggml スケジューラに自動登録され、`use_gpu=false` でも CPU 実行が 867ms → 3088ms に劣化する。Vulkan feature は Adreno 以外の GPU (Samsung RDNA 等) でのみ有効化すべき。
